satellite/satellitedb/certdb.go

```diff
@@ -19,82 +22,96 @@ type certDB struct {
 	db *dbx.DB
 }
 
-func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
+func (certs *certDB) SavePublicKey(ctx context.Context, nodeID storj.NodeID, pi *identity.PeerIdentity) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
-	}
 
-	_, err = certs.db.Get_CertRecord_By_Publickey(ctx, dbx.CertRecord_Publickey(pubbytes))
+	tx, err := certs.db.Begin()
 	if err != nil {
-		if err == sql.ErrNoRows {
-			return certs.tryAddPublicKey(ctx, nodeID, publicKey)
-		}
 		return Error.Wrap(err)
 	}
 
-	// publickey for the nodeID entry already exists, just return
-	return nil
-}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
 
-func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
-	defer mon.Task()(&ctx)(&err)
-	// no rows err, so create/insert an entry
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
+	if pi == nil {
+		return Error.New("Peer Identity cannot be nil")
 	}
+	chain := encodePeerIdentity(pi)
 
-	// TODO: use upsert here instead of create
-	_, err = certs.db.Create_CertRecord(ctx,
-		dbx.CertRecord_Publickey(pubbytes),
-		dbx.CertRecord_Id(nodeID.Bytes()),
-	)
-	// another goroutine might race to create the cert record, let's ignore that error
-	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
-		return nil
-	} else if err != nil {
+	var id int64
+	query := `SELECT id FROM certRecords WHERE peer_identity = ?;`
+	err = tx.QueryRow(certs.db.Rebind(query), chain).Scan(&id)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// create a new entry
+			_, err = tx.Exec(certs.db.Rebind(`INSERT INTO certRecords ( peer_identity, node_id, update_at ) VALUES ( ?, ?, ? );`), chain, nodeID.Bytes(), time.Now())
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			return nil
+		}
 		return Error.Wrap(err)
 	}
 
+	// already public key exists, just return
 	return nil
 }
 
 // GetPublicKey gets the public key of uplink corresponding to uplink id
-func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ crypto.PublicKey, err error) {
+func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ *identity.PeerIdentity, err error) {
 	defer mon.Task()(&ctx)(&err)
-	dbxInfo, err := certs.db.All_CertRecord_By_Id_OrderBy_Desc_UpdateAt(ctx, dbx.CertRecord_Id(nodeID.Bytes()))
+	dbxInfo, err := certs.db.All_CertRecord_By_NodeId_OrderBy_Desc_UpdateAt(ctx, dbx.CertRecord_NodeId(nodeID.Bytes()))
```

Comment:
Why use `All` when there can only be one answer?

Reply:
fixed